domains/client/domain-executor/src/gossip_message_validator.rs

```diff
@@ -211,19 +211,44 @@ where
         domain_id: DomainId,
         at: Block::Hash,
     ) -> Result<(), GossipMessageError> {
-        for extrinsic in extrinsics {
+        let block_number = self
+            .client
+            .number(at)?
+            .ok_or_else(|| sp_blockchain::Error::Backend(format!("Header for #{at} not found")))?;
+
+        for (index, extrinsic) in extrinsics.iter().enumerate() {
             let tx_hash = self.transaction_pool.hash_of(extrinsic);
 
             if self.transaction_pool.ready_transaction(&tx_hash).is_some() {
                 // TODO: Set the status of each tx in the bundle to seen
-            } else if let Err(_err) = self
+            } else if let Err(transaction_fee_err) = self
                 .client
                 .runtime_api()
                 .check_transaction_fee(at, extrinsic.clone())?
```

Comment:
Is it possible that the malicious node does not start the gossip worker and just submits the bundle directly to the primary chain, thus bypassing this check?

Reply:
Yes, it is. I had the same question, my understanding is that it won't comprise the security as the invalid transaction will be detected as an invalid state transition in that case. This path is an enhancement to the security and is intended to enforce the basic transaction legality even before the bundle is included in a primary block. Once we work on the compact bundle, i.e., the bundle gossip can be enabled, we may have more thinking and do something about this.

Reply:
> is intended to enforce the basic transaction legality even before the bundle is included in a primary block

Even if we can detect the illegal transaction earlier, it does not help much to prevent the bad bundle from including in the primary chain. Because the primary chain can not detect the illegal transaction and the malicious node can submit the bundle directly and successfully to the primary chain.

Also, we need the bad bundle to be included in the primary chain in order to verify the proof to ensure the proof is indeed targeting an existing bad bundle. Or we may need to include the whole bad bundle in the proof and verify the bundle signature to ensure the bad bundle is indeed produced by a legit executor.

Reply:
Preventing the bad bundle from being included in the primary chain is never a design goal for the illegal transaction proof. The goal is to find out this misbehavior and slash the malicious executor even before the bundle is able to be included in some primary block, the function of illegal transaction proof does not require the bundle to be included in the primary chain.

This reminds me that it's wrong to use the domain extrinsic builder here (this is likely the reason for causing these confusions, my bad :P). Basically, I misused the extrinsic index to retrieve the corresponding illegal transaction, the index shipped in the illegal transaction proof is **the index the extrinsics within the bundle not the index in the final domain block extrinsics**. Although I feel there is a chance to make it more efficient( perhaps the extrinsic as well as its merkle proof plus the bundle header is enough with some breaking changes), it's okay for me to include the entire bundle in the first iteration. You're right on verifying the bundle election, I missed that part as the slashing is still a TODO, but yes, we can do it in this PR. This will answer this comment https://github.com/subspace/subspace/pull/1404/files#r1178455251 too.

Reply:
Okay, make sense to me overall as the invalid transaction proof is intended to be an enhancement to the security but an essential one.

> as the invalid transaction will be detected as an invalid state transition in that case

In this way, we may also need to slash the executor for the invalid state transitions.